### DISCUSSION TITLE
Jetson Nano Driver Installation Issues (with possible solutions)

### DISCUSSION THREAD

**Discussion by abhisheksreesaila** [prl-mushr/mushr#39] [Software]
**Issue No 1**

The order of install specified in mushr_install.sh has to tweaked slightly I think.   

The current state : **ROS ==> MUSHR stack ==> HARDWARE**
- But the problem i ran into was MUSHR stack needed REAL SENSE SDK installed, hence the mushr stack CMAKE build did not complete successfully. 

**Solution (that worked for me)**

The correct order should be : **ROS ==>  real sense SDK ==> MUSHR ==> HARDWARE**

----------------------------------------------------------------------------------------------------

**Issue No 2**

When you flash a fresh image of JETPACK today, it will install Jetpack 4.5.2, but hang on

Jetpack 4.5.2 == supports CUDA 10.2 ==> which supports only torch 1.5 and above
But mushr_rhc install needs torch 1.1.  we can upgrade to torch 1.5, however,  torch 1.5 supports only python 3.6 which mushr does not support. So, 

**Solution (that worked for me)**

- Install torch 1.4 (the last torch version which supports python 2.7)
- Install "libopenblas" library [here](https://www.openblas.net/)
       - update bashrc as shown [here](https://www.programmersought.com/article/90674363798/)

** EASIER Solution **

Install Jetpack 4.2.2 from the NVIDIA archives to ease the pain of installation.

----------------------------------------------------------------------------------------------------

**Comment by schmittlema** [comment]
Hey @abhisheksreesaila. Thanks for the feedback. I think issue 1 was touched on in [this discussion](https://github.com/prl-mushr/mushr/discussions/37). Please let me know otherwise.

As for issue 2, yes this is a big pain point for us. In particular because NVIDIA has decided to not officially support (ever) Ubuntu 20.04 on the nano which supports ROS Noetic (python3). Jetpack 4.2 is the way to go and we recently updated our instructions in [this commit ](https://github.com/prl-mushr/mushr-website/commit/001966925f475a06baa1d021f8ac8a83bae354fa) to reflect that.

Thanks again for sharing what you found!

**Reply by abhisheksreesaila** [reply]
@schmittlema  From what i understand, the issue linked in your comment above deals with the proper commands to follow for real sense SDK to be installed.   What i am suggesting is just an ORDER of EXEC.   The current order of execution is  ROS stack ==> MUSHR stack ==> HARDWARE.  But thats didn't work for me since MUSHR stack depended on REAL SENSE to be installed first.  I had to first install REAL SENSE and come back to MUSHR STACK and compile for it to work.
**So I would recommend, ROS ==> REAL SENSE SDK ==> MUSHR STACK ==> REST of the hardware.**

**Comment by palanc** [comment]
Hi @abhisheksreesaila - sorry I had originally told @schmittlema that this problem was covered by the discussion he referenced, I should have read your original post more carefully. Although I think the script does eventually get everything set up correctly because it tries to compile the workspace multiple times, you're right that one of the compilation attempts fail. So I moved the installation of librealsense to before that point in our latest commit, so it should be good to go now. Thanks for letting us know!

**Reply by abhisheksreesaila** [reply]
awesome! glad it helped.
